team-telnyx/telnyx-webrtc-ios
# WEBRTC-672 - Push notification flow + CallKit

[WEBRTC-672 - Push notification flow](https://telnyx.atlassian.net/browse/WEBRTC-672)
---

Fix the Push notification flow to use CallKit. 

**Full PN flow:** 
1. User A login to B2BUA. We are including the following information inside the Login message:  `push_provider` / `push_token` / `push_environment` variables.
2. `User B` calls to `User A`. 
3. `User A` is connected to B2BUA so it will receive an INVITE message (There's no Push notification at this point)
4. `User A` kills the app and get's disconnected from B2BUA.
5. B2BUA stores the  `push_provider` / `push_token` / `push_environment`  for that user. 
6. `User B` calls `User A`
7. B2BUA detects that `User A` is not connected, then it parks the call and a PN is send to `User A`
8. `User A` receives the PN, registers a new call into **CallKit** (This will produce the native iOS Incoming call UI)and triggers in background the connection process to B2BUA
9. Once the connection process ends, B2BUA detects that there's a parked call for `User A` and sends the **INVITE** message.
10. The SDK will propagate the `onPushCall` event when the INVITE message is received.
11. `User A` can answer / reject the call from the Push notification. At this point, steps 9 and 10 should finish before the Answer / reject button is pressed by the user (This is for the POC)


## :older_man: :baby: Behaviors
### Before changes
- PN using CallKit was not supported

### After changes
- PN using CallKit is working

## TODO
- Handle possible errors during the re-connection process.

## ✋ Manual testing
On dev environment: 
1. Login with `User A` from mobile
2. Login with `User B` from the web dialer.
3. Kill the app for `User A`. 
4. Place a call from `User B` to `User A`
5. A push notification should be received on `User A` phone
6. Answer / Reject the call 
7. Call should be connected or rejected.
